### ISSUE TITLE
Authenticate with a thing

### ISSUE THREAD

**Issue by benfrancis** [WebThingsIO/android-app#21]
As a user I want to access a thing which is protected by authentication, e.g a username and password so that I can access things whilst protecting them from access by others.

Implementation of https://w3c.github.io/wot-thing-description/#sec-security-vocabulary-definition

**Comment by arist0v** [comment]
my first idea on this was to work with api key, i already have a working code that use the api key genereated from the gateway gui

**Comment by benfrancis** [comment]
Yes using the Bearer token (API key) is the right thing to do, but the important question is how the user gets the API key. We don't want them to have to type it in manually.

WebThings Gateway supports OAuth2 which should allow the user to authenticate with their username and password and be automatically issued an API key.

Here is an example section of a Thing Description which describes the OAuth2 endpoints.

```
"securityDefinitions": {
        "oauth2_sc": {
            "scheme": "oauth2",
            "flow": "code",
            "authorization": "https://plugfest.webthings.io/oauth/authorize",
            "token": "https://plugfest.webthings.io/oauth/token",
            "scopes": [
                "/things/virtual-things-6:readwrite",
                "/things/virtual-things-6",
                "/things:readwrite",
                "/things"
            ]
        }
    },
```
This metadata follows the OAuth2 security scheme defined in the WoT Thing Description specification 
https://www.w3.org/TR/wot-thing-description/#oauth2securityscheme

Eventually it would be good to support other security schemes too, but we should start with the OAuth2 scheme used by WebThings Gateway.

**Comment by arist0v** [comment]
Does it is a permenant api key? or a temporary that will need to be relogged with usernam and password each time? since my first plan was to mad a custom dashboard for old phone or tablet to be use as controller, i was going with a generated api key so once created i wont be bothered anymore so is the oauth_2 allow to request a permenant one?

**Comment by benfrancis** [comment]
As far as I know the tokens don't currently expire after a fixed period of time.

**Comment by arist0v** [comment]
Nice i will work the auth component with oauth 2 instead of direct token
